src/Workspaces/Remote/Core/AbstractAssetProviderExtensions.AnalyzerReference.cs

```diff
@@ -0,0 +1,118 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Serialization;
+using Roslyn.Utilities;
+
+#if NET
+using System.Runtime.Loader;
+#endif
+
+namespace Microsoft.CodeAnalysis.Remote;
+
+internal static partial class AbstractAssetProviderExtensions
+{
+#if NET
+
+    /// <summary>
+    /// Gate around <see cref="s_checksumToReferenceSet"/> to ensure it is only accessed and updated atomically.
+    /// </summary>
+    private static readonly SemaphoreSlim s_isolatedReferenceSetGate = new(initialCount: 1);
+
+    /// <summary>
+    /// Mapping from checksum for a particular set of assembly references, to the dedicated ALC and actual assembly
+    /// references corresponding to it.  As long as it is alive, we will try to reuse what is in memory.  But once it is
+    /// dropped from memory, we'll clean things up and produce a new one.
+    /// </summary>
+    private static readonly Dictionary<Checksum, WeakReference<IsolatedAssemblyReferenceSet>> s_checksumToReferenceSet = new();
+
+#endif
+
+    /// <summary>
+    /// Given a checksum for a set of analyzer references, fetches the existing ALC-isolated set of them if already
+    /// present in this process.  Otherwise, this fetches the raw serialized analyzer references from the host side,
+    /// then creates and caches an isolated set on the OOP side to hold onto them, passing out that isolated set of
+    /// references to be used by the caller (normally to be stored in a solution snapshot).
+    /// </summary>
+    public static async ValueTask<ImmutableArray<AnalyzerReference>> CreateIsolatedAnalyzerReferencesAsync(
+        this AbstractAssetProvider assetProvider,
+        AssetPath assetPath,
+        ChecksumCollection analyzerReferencesChecksum,
+        IAnalyzerAssemblyLoaderProvider assemblyLoaderProvider,
+        CancellationToken cancellationToken)
+    {
+        var checksum = analyzerReferencesChecksum.Checksum;
```

Comment:
>checksum [](http://example.com/codeflow?start=12&length=8)

nit: looks like def could be moved inside netcore specific directive.

Reply:
done.